tests/test_2306_cppyy_git.py

```diff
@@ -0,0 +1,118 @@
+
+
+import pytest
+import setuptools
+
+import awkward as ak
+import awkward._connect.cling
+
+cppyy = pytest.importorskip("cppyy")
+
+
+def test_array_as_generated_dataset():
+    array = ak.Array(
+        [
+            [{"x": 1, "y": [1.1]}, {"x": 2, "y": [2.2, 0.2]}],
+            [],
+            [{"x": 3, "y": [3.0, 0.3, 3.3]}],
+        ]
+    )
+
+    generator = ak._connect.cling.togenerator(array.layout.form, flatlist_as_rvec=False)
+    lookup = ak._lookup.Lookup(array.layout)
+
+    source_code = f"""
+    double go_fast(ssize_t length, ssize_t* ptrs) {{
+        auto awkward_array = {generator.dataset()};
+        double out = 0.0;
+
+        for (auto list : awkward_array) {{
+            for (auto record : list) {{
+                for (auto item : record.y()) {{
+                    out += item;
+                }}
+            }}
+        }}
+
+        return out;
+    }}
+    """
+
+    generator.generate(cppyy.cppdef)
+    cppyy.cppdef(source_code)
+    out = cppyy.gbl.go_fast(len(array), lookup.arrayptrs)
+    assert out == ak.sum(array["y"])
+
+
+@pytest.mark.skipif(
+    setuptools.extern.packaging.version.parse(cppyy.__version__)
+    < setuptools.extern.packaging.version.parse("3.0.1"),
+    reason="Awkward Array can only work with cppyy 3.0.1 or later.",
+)
+def test_array_as_type():
+    array = ak.Array(
+        [
+            [{"x": 1, "y": [1.1]}, {"x": 2, "y": [2.2, 0.2]}],
+            [],
+            [{"x": 3, "y": [3.0, 0.3, 3.3]}],
+        ]
+    )
+
+    source_code_cpp = f"""
+    double go_fast_cpp({array.cpptype} awkward_array) {{
+        double out = 0.0;
+
+        for (auto list : awkward_array) {{
+            for (auto record : list) {{
+                for (auto item : record.y()) {{
+                    out += item;
+                }}
+            }}
+        }}
+
+        return out;
+    }}
+    """
+
+    cppyy.cppdef(source_code_cpp)
+
+    out = cppyy.gbl.go_fast_cpp(array)
+    assert out == ak.sum(array["y"])
+
+
+@pytest.mark.skipif(
+    setuptools.extern.packaging.version.parse(cppyy.__version__)
+    < setuptools.extern.packaging.version.parse("3.0.1"),
+    reason="Awkward Array can only work with cppyy 3.0.1 or later.",
+)
+def test_array_as_templated_type():
+    array = ak.Array(
+        [
+            [{"x": 1, "y": [1.1]}, {"x": 2, "y": [2.2, 0.2]}],
+            [],
+            [{"x": 3, "y": [3.0, 0.3, 3.3]}],
+        ]
+    )
+
+    source_code_cpp = """
+    template<typename T>
+    double go_fast_cpp_2(T& awkward_array) {
+        double out = 0.0;
+
+        for (auto list : awkward_array) {
+            for (auto record : list) {
+                for (auto item : record.y()) {
+                    out += item;
+                }
+            }
+        }
+
+        return out;
+    }
+    """
+
+    cppyy.cppdef(source_code_cpp)
+
+    out = cppyy.gbl.go_fast_cpp_2[array.cpptype](array)
```

Comment:
This is great! The fact that users can write functions with this interface really streamlines it. Thanks, @wlav, for cppyy 3.0.1!

I still have the question (for @ianna) of whether the same array twice or a different array with the same `Form` (and therefore the same `array.cpptype`, including the classes that get defined in the evaluation of that property) would initiate a new compilation. Another call of this function with the same-typed (same `Form`) array _should_ be able to reuse precompiled code, and that becomes important for a Dask worker evaluating partitions of an array.

Reply:
> This is great! The fact that users can write functions with this interface really streamlines it. Thanks, @wlav, for cppyy 3.0.1!
> 
> I still have the question (for @ianna) of whether the same array twice or a different array with the same `Form` (and therefore the same `array.cpptype`, including the classes that get defined in the evaluation of that property) would initiate a new compilation. Another call of this function with the same-typed (same `Form`) array _should_ be able to reuse precompiled code, and that becomes important for a Dask worker evaluating partitions of an array.

@jpivarski  - by default the generated C++ code is cached and compiled only once - thanks to you - yes, you've thought about it before and implemented it here:
https://github.com/scikit-hep/awkward/blob/main/src/awkward/_connect/cling.py#L47#L57